src/leetCodeExplorer.ts

```diff
@@ -7,6 +7,8 @@ import * as list from "./commands/list";
 import { leetCodeManager } from "./leetCodeManager";
 import { ProblemState } from "./shared";
 
+export type Category = "Difficulty" | "Tag" | "Company";
```

Comment:
Can we use enum instead?

Reply:
Have you tried this:

```ts
export enum Category {
    Difficulty = "Difficulty",
    Tag = "Tag",
    Company = "Company",
}
```

Reply:
No need to call `.toString()` here.

Whenever you want to use the Category value as a string, just using `Category.xxxx`.

By doing this, you will get the value as a string meanwhile keep the typedef.

I tried using enum here and everything works fine.

Reply:
Let's make the `Category` enum type:
```typescript
export enum Category {
    Difficulty = "Difficulty",
    Tag = "Tag",
    Company = "Company",
}
```

Then rewrite [line 147](https://github.com/jdneo/vscode-leetcode/blob/d9abd5f68a92bc7679066ea35bd342fd80a599ea/src/leetCodeExplorer.ts#L147) this way:
```typescript
const categories: Map<Category, string[]> = new Map<Category, string[]>([
    [Category.Difficulty, [problem.difficulty]],
    [Category.Tag, problem.tags],
    [Category.Company, problem.companies],
]);
```

Follows the iteration:
```typescript
for (const [parent, children] of categories) {
    for (let subCategory of children) {
```
Here, `typeof parent` is `Category`:
![image](https://user-images.githubusercontent.com/20227484/52049503-fbb16e00-2588-11e9-8ce3-3c3159ba6445.png)


Continues the [line 156](https://github.com/jdneo/vscode-leetcode/blob/d9abd5f68a92bc7679066ea35bd342fd80a599ea/src/leetCodeExplorer.ts#L156):
```typescript
const problems: list.IProblem[] | undefined = this.treeData[parent].get(subCategory);
```

Since `typeof parent` is `Category` and `this.treeData`'s key type is `string`, `get`'s type could not be deduced from the usage, and there's no way to use `Category.xxxx` here:
![image](https://user-images.githubusercontent.com/20227484/52049499-f522f680-2588-11e9-86e8-ed980f86da31.png)

Meanwhile, we couldn't change `this.treeData` type to `Map<Category, Map<string, list.IProblem[]>> since there is a field `Favorite` which doesn't processed as `Category` and its value type is `list.IProblem[]`.

If you insist to use enum here, a workaround could be:
```typescript
private treeData: {
    [Category.Difficulty]: Map<string, list.IProblem[]>,
    [Category.Tag]: Map<string, list.IProblem[]>,
    [Category.Company]: Map<string, list.IProblem[]>,
    Favorite: list.IProblem[],
};
```

Now `get`'s type could be deduced properly:
![image](https://user-images.githubusercontent.com/20227484/52049958-10dacc80-258a-11e9-8416-6f80f5ef9392.png)

Yet I think it complicates what could have been an easy and elegant solution in Typescript.

Reply:
We need to write more hard-coded strings if using string union here while using enum can get rid of this. That's the reason I prefer using enum here: make the code easier to maintain in the future.

There are many ways to get avoid the special field `Favorite`. One simple way to do a simple check before getting the map. TS will help us to infer the type during compilation. See my code for more details.